Orchestrator/Prognetics.Saga.Orchestrator/SagaEngine.cs

```diff
@@ -1,33 +1,125 @@
+using Microsoft.Extensions.Logging;
 using Prognetics.Saga.Core.Abstract;
+using Prognetics.Saga.Core.Model;
 using Prognetics.Saga.Orchestrator.Contract;
 using Prognetics.Saga.Orchestrator.Contract.DTO;
+using System.Transactions;
 
 namespace Prognetics.Saga.Orchestrator;
 
 public class SagaEngine : ISagaEngine
 {
+    private readonly ISagaLog _sagaLog;
+    private readonly ICompensationStore _compensationStore;
     private readonly ITransactionLedgerAccessor _transactionLedgerAccessor;
+    private readonly IdGenerator _idGenerator;
+    private readonly ILogger<ISagaEngine> _logger;
 
-    public SagaEngine(ITransactionLedgerAccessor transactionLedgerAccessor)
+    public SagaEngine(
+        ISagaLog sagaLog,
+        ICompensationStore compensationStore,
+        ITransactionLedgerAccessor transactionLedgerAccessor,
+        IdGenerator idGenerator,
+        ILogger<ISagaEngine> logger)
     {
+        _sagaLog = sagaLog;
+        _compensationStore = compensationStore;
         _transactionLedgerAccessor = transactionLedgerAccessor;
+        _idGenerator = idGenerator;
+        _logger = logger;
     }
 
-    public Task<EngineOutput?> Process(EngineInput input)
+    public async Task<EngineOutput?> ProcessOrDefault(EngineInput input)
     {
         var transactionLedger = _transactionLedgerAccessor.TransactionsLedger;
-        var transactionModel = transactionLedger.GetTransactionByCompletionEventName(input.EventName);
-        var stepRecord = transactionModel?.GetStepByCompletionEventName(input.EventName);
+        var transactionStep = transactionLedger.GetTransactionStepByCompletionEventName(input.EventName);
 
-        if (stepRecord is null || transactionModel is null)
+        if (transactionStep is null)
         {
-            return Task.FromResult((EngineOutput?)null);
+            _logger.LogError("Unknown event name: {EventName}", input.EventName);
+            return null;
         }
 
-        return Task.FromResult((EngineOutput?)new EngineOutput(
-            stepRecord.Step.CompletionEventName,
-            new OutputMessage(
-                input.Message.TransactionId ?? Guid.NewGuid().ToString(),
-                input.Message.Payload)));
+        var transactionId = input.Message.TransactionId;
+
+        if (transactionStep.IsFirst)
+        {
+            transactionId = await InitializeTransaction(transactionStep);
+        }
+        else if (transactionId is null)
+        {
+            _logger.LogError("Transaction id is missing");
+            return null;
```

Comment:
What does it mean if we return null? Didn't we want to avoid using nulls?

Reply:
I see a few ways of dealing with invalid messages:
- returning null if processing have failed, the method name has a postfix 'OrDefault'
- returning bool and the engine output as an out parameter
- throwing an exception
- IsSuccess prop in the engine output
I've chosen the first one because I like dealing with nulls with the nullable feature enabled in .NET. However i'm open to the other solutions as well.

Reply:
As discussed let's go with the engine output :)